### ISSUE TITLE
Is it possible to disallow the creation of a "dumb-terminal"?

### ISSUE THREAD

**Issue by Osiris-Team** [jline/jline3#671]
My situation:
I got 2 applications (executable jar files), let's call them MainApp and SecondaryApp.
Both apps have a version of jline (with jansi) installed.

Workflow:
I start the MainApp via a script (I'm on windows btw) and then the SecondaryApp programmatically from within the MainApp (that's shown in the code below).

My issues:
```java
ProcessBuilder processBuilder = new ProcessBuilder(commands);

//processBuilder.inheritIO(); // This works fine, the output is colored,
// but the user input is kind of broken: Input gets shown in the terminal,
// only when pressing and holding down the key.
// Not using inheritIO() will cause the creation of a dumb-terminal.
// In this case the output won't be colored, but user input will work fine.
// My problem could be solved, if there would be a way of getting the SecondaryApps ansi outputstream.
// Or by forcing the creation of a "not-dumb-terminal".
// Any ideas how to achieve that? Or any alternatives?

processSecondaryApp = processBuilder.start();
```

**Comment by gnodet** [comment]
In order to let the second application use the terminal correctly, you need to make sure the first application is not using it.
This usually means resetting the terminal to its original attributes if you've changed those (the `ReadLine` restores it after reading a line, so that one should be safe), and also _pausing_ the terminal : some terminals need to start a thread that read the input and put them in the terminal's internal buffer.  This is certainly your problem.  Try calling `terminal.pause(true);` before starting your second application, and later `terminal.resume();` to start the input reading again.

**Comment by Osiris-Team** [comment]
Thanks I'll give it a try!

**Comment by Osiris-Team** [comment]
Yep this works. But after resuming the MainApps terminal the problem returns.
I guess there is no workaround for that :/

**Comment by Osiris-Team** [comment]
What about virtual terminals?

**Comment by gnodet** [comment]
> Yep this works. But after resuming the MainApps terminal the problem returns.
> I guess there is no workaround for that :/

Which problem returns ?

**Comment by gnodet** [comment]
> What about virtual terminals?

You can also

**Comment by gnodet** [comment]
If you're simply interested in the colored output, you can pass  `-Dorg.jline.terminal.dumb.color=true`to the forked JVM for the second app, this should allow the output to be coloured when using a dumb terminal.

**Comment by Osiris-Team** [comment]
> > Yep this works. But after resuming the MainApps terminal the problem returns.
> > I guess there is no workaround for that :/
> 
> Which problem returns ?

^The weird user input thing. Where when I press a key sometimes it won't recognize that I pressed it. Sometimes I need to hold the key down so it gets an input.

**Comment by Osiris-Team** [comment]
Ah ok I understood it.
So there is no way to have 2 line readers/scanners at the same time.
I will have to pause the MainApps terminal as long as the SecondaryApp is active.

**Comment by Osiris-Team** [comment]
@gnodet Hey it's me again, with another problem haha.
This is getting more and more complicated and I am slowly starting to lose my mind. So Its ok if you can't help me with this one.
Since the inheritIO() thing could be solved. Thanks again. Now I want to duplicate the terminals output and send it my online html console.
For that I am doing this:
```java
// Set default SysOut to TeeOutput, for the OnlineConsole
            AnsiConsole.systemInstall(); // This must happen before the stuff below.
            // Else the pipedOut won't display ansi. Idk why though...
            PIPED_IN = new NonBlockingPipedInputStream();
            OutputStream pipedOut = new PipedOutputStream(PIPED_IN);
            MyTeeOutputStream teeOut = new MyTeeOutputStream(TERMINAL.output(), pipedOut);
            PrintStream newOut = new PrintStream(teeOut);
            System.setOut(newOut); // This causes
            // the standard System.out stream to be mirrored to pipedOut, which then can get
            // read by PIPED_IN. This ensures, that the original System.out is not touched.
```
The code above, is executed in the MainApp. After that the SecondaryApp is started with inheritIO().
Until here everything works fine. The SecondaryApps colors, user input and output are working.
The problem is that its output is cannot be read by my custom `PIPED_IN`.
I couldn't figure out why.

Here is the MyTeeOutputStream class: https://github.com/Osiris-Team/AutoPlug-Client/blob/master/src/main/java/com/osiris/autoplug/client/utils/MyTeeOutputStream.java
Here is the NonBlockingPipedInputStream class: https://github.com/Osiris-Team/AutoPlug-Client/blob/master/src/main/java/com/osiris/autoplug/client/utils/NonBlockingPipedInputStream.java

**Comment by Osiris-Team** [comment]
> If you're simply interested in the colored output, you can pass `-Dorg.jline.terminal.dumb.color=true`to the forked JVM for the second app, this should allow the output to be coloured when using a dumb terminal.

@gnodet this didnt work for me. Is there an alternative, or is there a list of all jline jvm arguments somewhere?

**Comment by gnodet** [comment]
> > If you're simply interested in the colored output, you can pass `-Dorg.jline.terminal.dumb.color=true`to the forked JVM for the second app, this should allow the output to be coloured when using a dumb terminal.
> 
> @gnodet this didnt work for me. Is there an alternative, or is there a list of all jline jvm arguments somewhere?

https://github.com/jline/jline3/blob/master/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java#L46-L53 ?

**Comment by Osiris-Team** [comment]
@gnodet Tried all of them:
`` java
commands.add("-Dorg.jline.terminal.dumb=true");
commands.add("-Dorg.jline.terminal.dumb.color=true");
commands.add("-Dorg.jline.terminal.type=dumb-color");
```
without success,
could it be that the child application is overwriting those values?
